File: ai/research/pattern_alignment_risk_governor.py
"""
Multi-Timeframe Harmonic & Geometric Pattern Alignment Governor (Module BD1 - Python)
Synthesizes "Trade Chart Patterns Guide":
- Risk-to-Reward Ratio Auditor: RR = |Target - Entry| / |Stop - Entry| >= 2.0
- Higher-Timeframe (HTF) vs Lower-Timeframe (LTF) Geometric Trend Confluence
- Automated Pattern Invalidation Governor
"""

from typing import Dict, List, Any


class PatternAlignmentRiskGovernor:
    def __init__(self, min_rr_ratio: float = 2.0):
        self.min_rr = min_rr_ratio

    def audit_pattern_risk_reward(
        self,
        entry_price: float,
        target_price: float,
        stop_loss_price: float,
        htf_trend_direction: int,  # +1 Bull, -1 Bear, 0 Neutral
        pattern_direction: int     # +1 Bull, -1 Bear
    ) -> Dict[str, Any]:
        """
        Audits setup against institutional risk rules and multi-timeframe confluence.
        """
        reward = abs(target_price - entry_price)
        risk = abs(entry_price - stop_loss_price)
        rr_ratio = reward / max(1e-4, risk)

        is_rr_approved = rr_ratio >= self.min_rr
        is_htf_aligned = (htf_trend_direction == pattern_direction) or (htf_trend_direction == 0)

        is_trade_approved = is_rr_approved and is_htf_aligned

        return {
            "entry_price": entry_price,
            "target_price": target_price,
            "stop_loss_price": stop_loss_price,
            "reward_points": round(reward, 2),
            "risk_points": round(risk, 2),
            "rr_ratio": round(rr_ratio, 2),
            "is_rr_approved": is_rr_approved,
            "is_htf_aligned": is_htf_aligned,
            "is_trade_approved": is_trade_approved,
            "verdict": "APPROVED_EXECUTE" if is_trade_approved else "BLOCKED_SUBOPTIMAL_SETUP"
        }

File: engine/java/src/main/java/com/optionalpha/research/PatternAlignmentRiskGovernor.java
package com.optionalpha.research;

import java.util.HashMap;
import java.util.Map;

/**
 * Module BD2 (Java): Multi-Timeframe Harmonic & Geometric Pattern Alignment Governor.
 * Enforces R:R >= 2.0 and multi-timeframe directional trend confluence in Java.
 */
public class PatternAlignmentRiskGovernor {
    private final double minRr;

    public PatternAlignmentRiskGovernor(double minRr) {
        this.minRr = minRr;
    }

    public PatternAlignmentRiskGovernor() {
        this(2.0);
    }

    public Map<String, Object> auditSetup(double entry, double target, double stopLoss, int htfDir, int patternDir) {
        Map<String, Object> res = new HashMap<>();
        double reward = Math.abs(target - entry);
        double risk = Math.abs(entry - stopLoss);
        double rr = reward / Math.max(1e-4, risk);

        boolean rrOk = rr >= this.minRr;
        boolean htfOk = (htfDir == patternDir) || (htfDir == 0);
        boolean approved = rrOk && htfOk;

        res.put("rrRatio", rr);
        res.put("isApproved", approved);
        res.put("verdict", approved ? "APPROVED_EXECUTE" : "BLOCKED");
        return res;
    }
}
